<a href="https://colab.research.google.com/github/vhpraneeth/VIT_DWDM/blob/main/Lab_9_DWDM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

18BCE7147

[GitHub Link](https://github.com/vhpraneeth/VIT_DWDM/blob/main/Lab_9_DWDM.ipynb)

## Ensemble Methods 


In [2]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import pandas as pd

In [3]:
df = sns.load_dataset('titanic')
df.dropna(inplace=True)

In [4]:
from sklearn import preprocessing
lb = preprocessing.LabelBinarizer()

In [5]:
X = df[['pclass', 'sex', 'age']]
X['sex'] = lb.fit_transform(X['sex'])
y = df['survived']

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [6]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

# Boosting 



In [7]:
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

def print_score(clf, X_train, y_train, X_test, y_test, train=True):
    'print the accuracy score, classification report and confusion matrix of classifier'
    if train:
        'training performance'
        print("Train Result:\n")
        print("accuracy score: {0:.4f}\n".format(accuracy_score(y_train, clf.predict(X_train))))
        print("Classification Report: \n {}\n".format(classification_report(y_train, clf.predict(X_train))))
        print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_train, clf.predict(X_train))))

        res = cross_val_score(clf, X_train, y_train, cv=10, scoring='accuracy')
        print("Average Accuracy: \t {0:.4f}".format(np.mean(res)))
        print("Accuracy SD: \t\t {0:.4f}".format(np.std(res)))
        
    elif train==False:
        'test performance'
        print("Test Result:\n")        
        print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, clf.predict(X_test))))
        print("Classification Report: \n {}\n".format(classification_report(y_test, clf.predict(X_test))))
        print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, clf.predict(X_test))))    
        

In [8]:
from sklearn import preprocessing
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix, roc_auc_score
def print_score(clf, X_train, X_test, y_train, y_test, train=True):
    '''
    v0.1 Follow the scikit learn library format in terms of input
    print the accuracy score, classification report and confusion matrix of classifier
    '''
    lb = preprocessing.LabelBinarizer()
    lb.fit(y_train)
    if train:
        'training performance'
        res = clf.predict(X_train)
        print("Train Result:\n")
        print("accuracy score: {0:.4f}\n".format(accuracy_score(y_train, 
                                                                res)))
        print("Classification Report: \n {}\n".format(classification_report(y_train, 
                                                                            res)))
        print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_train, 
                                                                  res)))
        print("ROC AUC: {0:.4f}\n".format(roc_auc_score(lb.transform(y_train), 
                                                      lb.transform(res))))

        res = cross_val_score(clf, X_train, y_train, cv=10, scoring='accuracy')
        print("Average Accuracy: \t {0:.4f}".format(np.mean(res)))
        print("Accuracy SD: \t\t {0:.4f}".format(np.std(res)))
        
    elif train==False:
        'test performance'
        res_test = clf.predict(X_test)
        print("Test Result:\n")        
        print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, 
                                                                res_test)))
        print("Classification Report: \n {}\n".format(classification_report(y_test, 
                                                                            res_test)))
        print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, 
                                                                  res_test)))   
        print("ROC AUC: {0:.4f}\n".format(roc_auc_score(lb.transform(y_test), 
                                                      lb.transform(res_test))))
        

# AdaBoost / Adaptive Boosting





In [9]:
from sklearn.ensemble import AdaBoostClassifier
ada_clf = AdaBoostClassifier()
ada_clf.fit(X_train, y_train)

AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None, learning_rate=1.0,
                   n_estimators=50, random_state=None)

In [10]:
print_score(ada_clf, X_train, X_test, y_train, y_test, train=True)

Train Result:

accuracy score: 0.9134

Classification Report: 
               precision    recall  f1-score   support

           0       0.89      0.82      0.85        38
           1       0.92      0.96      0.94        89

    accuracy                           0.91       127
   macro avg       0.90      0.89      0.89       127
weighted avg       0.91      0.91      0.91       127


Confusion Matrix: 
 [[31  7]
 [ 4 85]]

ROC AUC: 0.8854

Average Accuracy: 	 0.7654
Accuracy SD: 		 0.0879


In [11]:
print_score(ada_clf, X_train, X_test, y_train, y_test, train=False)

Test Result:

accuracy score: 0.7455

Classification Report: 
               precision    recall  f1-score   support

           0       0.68      0.62      0.65        21
           1       0.78      0.82      0.80        34

    accuracy                           0.75        55
   macro avg       0.73      0.72      0.73        55
weighted avg       0.74      0.75      0.74        55


Confusion Matrix: 
 [[13  8]
 [ 6 28]]

ROC AUC: 0.7213



## AdaBoost with Random Forest

In [12]:
from sklearn.ensemble import RandomForestClassifier
ada_clf = AdaBoostClassifier(RandomForestClassifier())
ada_clf.fit(X_train, y_train)

AdaBoostClassifier(algorithm='SAMME.R',
                   base_estimator=RandomForestClassifier(bootstrap=True,
                                                         ccp_alpha=0.0,
                                                         class_weight=None,
                                                         criterion='gini',
                                                         max_depth=None,
                                                         max_features='auto',
                                                         max_leaf_nodes=None,
                                                         max_samples=None,
                                                         min_impurity_decrease=0.0,
                                                         min_impurity_split=None,
                                                         min_samples_leaf=1,
                                                         min_samples_split=2,
                                       

In [13]:
print_score(ada_clf, X_train, X_test, y_train, y_test, train=True)

Train Result:

accuracy score: 0.9528

Classification Report: 
               precision    recall  f1-score   support

           0       0.97      0.87      0.92        38
           1       0.95      0.99      0.97        89

    accuracy                           0.95       127
   macro avg       0.96      0.93      0.94       127
weighted avg       0.95      0.95      0.95       127


Confusion Matrix: 
 [[33  5]
 [ 1 88]]

ROC AUC: 0.9286

Average Accuracy: 	 0.7231
Accuracy SD: 		 0.1475


In [14]:
print_score(ada_clf, X_train, X_test, y_train, y_test, train=False)

Test Result:

accuracy score: 0.7636

Classification Report: 
               precision    recall  f1-score   support

           0       0.79      0.52      0.63        21
           1       0.76      0.91      0.83        34

    accuracy                           0.76        55
   macro avg       0.77      0.72      0.73        55
weighted avg       0.77      0.76      0.75        55


Confusion Matrix: 
 [[11 10]
 [ 3 31]]

ROC AUC: 0.7178



# Gradient Boosting / Gradient Boosting Machine (GBM)

In [15]:
from sklearn.ensemble import GradientBoostingClassifier
gbc_clf = GradientBoostingClassifier()
gbc_clf.fit(X_train, y_train)

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='deprecated',
                           random_state=None, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [16]:
print_score(gbc_clf, X_train, X_test, y_train, y_test, train=True)

Train Result:

accuracy score: 0.9528

Classification Report: 
               precision    recall  f1-score   support

           0       0.97      0.87      0.92        38
           1       0.95      0.99      0.97        89

    accuracy                           0.95       127
   macro avg       0.96      0.93      0.94       127
weighted avg       0.95      0.95      0.95       127


Confusion Matrix: 
 [[33  5]
 [ 1 88]]

ROC AUC: 0.9286

Average Accuracy: 	 0.7468
Accuracy SD: 		 0.1616


In [17]:
print_score(gbc_clf, X_train, X_test, y_train, y_test, train=False)

Test Result:

accuracy score: 0.8182

Classification Report: 
               precision    recall  f1-score   support

           0       0.87      0.62      0.72        21
           1       0.80      0.94      0.86        34

    accuracy                           0.82        55
   macro avg       0.83      0.78      0.79        55
weighted avg       0.83      0.82      0.81        55


Confusion Matrix: 
 [[13  8]
 [ 2 32]]

ROC AUC: 0.7801



| Classifier | Decision Tree | Bagging | Random Forest | Optimised RF | Extra-Trees | AdaBoost (CART) | AdaBoost (RF) | Gradient Boosting |
|:-|:-|:- |:- |:- |:- |:-|:-| :- |
| Train accuracy score | 0.9528 | 0.9528 | 0.9325 | 0.9264 | 0.9448 | 0.8661 | 0.9528 | 0.9449 |
| Average accuracy score | 0.7724 | 0.7879 | 0.7801 | 0.7059 | 0.7548 | 0.7793 | 0.7353 | 0.7906 |
| SD | 0.1018 | 0.1008 | 0.1474 | 0.1308 | 0.1406 | 0.1172 | 0.0881 | 0.0912 |
| Test accuracy score | 0.7636 | 0.7455 | 0.7895 | 0.6316 | 0.7895 | 0.6545 | 0.7818 | 0.7818 |


# XGBoost (Extreme Gradient Boosting)



In [18]:
import xgboost as xgb

In [19]:
xgb_clf = xgb.XGBClassifier(max_depth=5, n_estimators=10000, learning_rate=0.3,
                            n_jobs=-1)

In [20]:
xgb_clf.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.3, max_delta_step=0, max_depth=5,
              min_child_weight=1, missing=None, n_estimators=10000, n_jobs=-1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [21]:
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

def print_score(clf, X_train, y_train, X_test, y_test, train=True):
    'print the accuracy score, classification report and confusion matrix of classifier'
    if train:
        'training performance'
        print("Train Result:\n")
        print("accuracy score: {0:.4f}\n".format(accuracy_score(y_train, clf.predict(X_train))))
        print("Classification Report: \n {}\n".format(classification_report(y_train, clf.predict(X_train))))
        print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_train, clf.predict(X_train))))

        res = cross_val_score(clf, X_train, y_train, cv=10, scoring='accuracy')
        print("Average Accuracy: \t {0:.4f}".format(np.mean(res)))
        print("Accuracy SD: \t\t {0:.4f}".format(np.std(res)))
        
    elif train==False:
        'test performance'
        print("Test Result:\n")        
        print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, clf.predict(X_test))))
        print("Classification Report: \n {}\n".format(classification_report(y_test, clf.predict(X_test))))
        print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, clf.predict(X_test))))    
        

In [22]:
print_score(xgb_clf, X_train, y_train, X_test, y_test, train=True)

Train Result:

accuracy score: 0.9370

Classification Report: 
               precision    recall  f1-score   support

           0       0.97      0.82      0.89        38
           1       0.93      0.99      0.96        89

    accuracy                           0.94       127
   macro avg       0.95      0.90      0.92       127
weighted avg       0.94      0.94      0.94       127


Confusion Matrix: 
 [[31  7]
 [ 1 88]]

Average Accuracy: 	 0.7468
Accuracy SD: 		 0.1064


In [23]:
print_score(xgb_clf, X_train, y_train, X_test, y_test, train=False)

Test Result:

accuracy score: 0.8000

Classification Report: 
               precision    recall  f1-score   support

           0       0.86      0.57      0.69        21
           1       0.78      0.94      0.85        34

    accuracy                           0.80        55
   macro avg       0.82      0.76      0.77        55
weighted avg       0.81      0.80      0.79        55


Confusion Matrix: 
 [[12  9]
 [ 2 32]]

